### PR TITLE
KDF "info" parameters concatenate when multiple are given

### LINKED ISSUE #20717: OSSL_KDF_PARAM_INFO isn't consistent
Currently four KDFs accept the `OSSL_KDF_PARAM_INFO` parameter:

1. HKDF
2. KBKDF
3. SSKDF
4. X9.63 KDF

Of these, HKDF concatenates all of these parameters into a single lump.  The other three only accept one instance of the parameter and use that.  Ideally, all would behave consistently.  I.e. the way HKDF does.  This will make them all consistent, it will allow the parameter to be documented identically and it eases the end user's task by allowing them to write self documenting code that shows that the pieces are correctly concatenated.

With respect to the CAVP tests: HKDF and SSKDF require several things concatenated.  Thus is makes sense to, at least, extend SSKDF this way.

As for implementation, this is really just exposing the code that already exists in HKDF to the other three.

### PR DESCRIPTION
**Pull request by paulidale**
When multiple OSSL_KDF_PARAM_INFO parameters are passed to a ctx_set function, they are concatenated to form the final info.  This was already being done in HKDF and should have been done in SSKDF & X9.63 KDF.  The only remaining KDF that uses this parameter is KBKDF and for consistency, it was changed here too.

Also deduplicated the allocate memory and copy a parameter function into a single instance.

Fixes #20717

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by t-j-h** [comment]
nice cleanup - but I don't think the new utility param(s) functions should really be in the kdf area - they belong up with the param(s) implementation as they are not really KDF specific if you take the step of having the OSSL_KDF_PARAM_INFO as a parameter to the functions - i.e. it is straight forward to have these completely generic.

Or is this being kept deliberately isolated to enable potential backfit? And in which case I'd still suggest the right going-forward approach is generic (and in the PARAM(S) API itself.

**Comment by paulidale** [comment]
I deliberately limited this to KDFs, that's the scope of the issue.

A more generic solution is possible and PARAMs would be a good place I suspect.

**Comment by paulidale** [comment]
I will generalise the functions and make them available internally, I like that suggestion.  We can discuss making them public later if someone wants them.

As for back porting, we can discuss later.

**Comment by paulidale** [comment]
Moved & renamed the new functions and made them more widely available.

File: include/internal/params.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * Passing 0 for maxsize means unlimited size.
+ *
+ * Returns 1 on success, 0 on failure and -1 if there are no matching params.
+ *
+ * *out and *out_len are guaranteed to be untouched if this function
+ * doesn't return success.
+ */
+int ossl_param_get1_concat_octet_string(const OSSL_PARAM *params, const char *name,
+                                        unsigned char **out, size_t *size,
```
Comment (t8m) [inline include/internal/params.h:37]: You're talking about `out_len` instead of `size` in the comment above.

File: crypto/params.c [hunk 2]
```diff
@@ -1481,6 +1483,130 @@ OSSL_PARAM OSSL_PARAM_construct_octet_ptr(const char *key, void **buf,
...
+ * *out will point to an allocated buffer on successful return.
+ * Any existing allocation in *out is cleared and freed.
+ *
+ * Passing 0 for maxsize means unlimited size.
+ *
+ * Returns 1 on success, 0 on failure and -1 if there are no matching params.
+ *
+ * *out and *out_len are guaranteed to be untouched if this function
+ * doesn't return success.
+ */
```
Comment (t8m) [inline crypto/params.c:1565]: I wouldn't duplicate the description here if it's already in the header.

File: crypto/params.c [hunk 3]
```diff
@@ -1481,6 +1483,130 @@ OSSL_PARAM OSSL_PARAM_construct_octet_ptr(const char *key, void **buf,
...
 
+/*
+ * Extract the parameter into an allocated buffer.
+ * Any existing allocation in *out is cleared and freed.
+ *
+ * Returns 1 on success, 0 on failure and -1 if there are no matching params.
+ *
+ * *out and *out_len are guaranteed to be untouched if this function
+ * doesn't return success.
+ */
```
Comment (t8m) [inline crypto/params.c:1494]: Do not duplicate the description from header

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/params.c (hunk 2 above)
Comment (paulidale) [reply crypto/params.c:1565]: Better might be having the comment here.  Developers rarely look at the headers and jump straight to the code.

**Comment by paulidale** [comment]
Ping @openssl/committers for review.

**Review by t8m — APPROVED** [review]

**Comment by slontis** [comment]
Not sure I agree with this approach. It was  done in one place to be backwards compatible. How useful is this really?
Should we also limit the size?

**Review by hlandau — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks.

**Pull request closed without merging**